src/LeafInputSignature.cpp

```diff
@@ -0,0 +1,192 @@
+                        continue;
+                    HashSetInsert(&implicitLeafInputs, includedFile.m_FilenameHash, includedFile.m_Filename);
+                }
+            }
+        }
+    }
+
+    auto addFileContentsToHash = [&](const char* filename, uint32_t filename_hash, const char* label)
+    {
+        HashDigest digest = ComputeFileSignatureSha1(stat_cache, digest_cache, filename, filename_hash);
+
+        if (ingredient_stream)
+        {
+            char digestString[kDigestStringSize];
+            DigestToString(digestString, digest);
+            char buffer[1000];
+            strncpy(buffer, filename, sizeof(buffer));
+            char*p = buffer;
+            for ( ; *p; ++p) *p = tolower(*p);
+            fprintf(ingredient_stream, "%s: %s %s\n", label, digestString, buffer);
+        }
+        HashAddPath(&hashState, filename);
+        HashUpdate(&hashState, &digest, sizeof(digest));
+    };
+
+    HashSetWalk(&explicitLeafInputs, [&](uint32_t index, uint32_t hash, const char *filename) {
+        addFileContentsToHash(filename, hash, "explicitLeafInput");
+    });
+    HashSetWalk(&implicitLeafInputs, [&](uint32_t index, uint32_t hash, const char *filename) {
+        addFileContentsToHash(filename, hash, "implicitLeafInput");
+    });
+
+    HashDigest result;
+    HashFinalize(&hashState, &result);
+    return result;
+}
+
+//This function calculates the offline part of the signature, which we store in the dag-derived file
+static HashDigest CalculateLeafInputHashOffline(const Frozen::Dag* dag, int32_t nodeIndex, MemAllocHeap* heap, FILE* ingredient_stream)
+{
+    HashDigest hashResult = {};
+
+    Buffer<int32_t> all_dependent_nodes;
+    BufferInit(&all_dependent_nodes);
+
+    FindDependentNodesFromRootIndex(heap, dag, nodeIndex, all_dependent_nodes);
+
+    HashState hashState;
+    HashInit(&hashState);
+
+    for(int32_t childNodeIndex : all_dependent_nodes)
+    {
+        auto& dagNode = dag->m_DagNodes[childNodeIndex];
+
+        if (ingredient_stream)
+            fprintf(ingredient_stream, "\nannotation: %s\n", dagNode.m_Annotation.Get());
+
+        HashAddString(ingredient_stream, &hashState, "action", dagNode.m_Action.Get());
+
+        for(auto& env: dagNode.m_EnvVars)
+        {
+            HashAddString(ingredient_stream, &hashState, "env_name", env.m_Name);
+            HashAddString(ingredient_stream, &hashState, "env_value", env.m_Value);
+        }
+        for (auto& s: dagNode.m_AllowedOutputSubstrings)
+            HashAddString(ingredient_stream, &hashState, "allowed_outputstring", s);
+        for (auto& f: dagNode.m_OutputFiles)
+            HashAddString(ingredient_stream, &hashState, "output", f.m_Filename.Get());
+
+        int relevantFlags = dagNode.m_Flags & ~Frozen::DagNode::kFlagCacheableByLeafInputs;
+        if (relevantFlags != (Frozen::DagNode::kFlagOverwriteOutputs | Frozen::DagNode::kFlagAllowUnexpectedOutput))
```

Comment:
Why are we only writing flags if they are != (Frozen::DagNode::kFlagOverwriteOutputs | Frozen::DagNode::kFlagAllowUnexpectedOutput)?

Reply:
these are the default flags. omitting them makes the ingredient stream smaller.